substrate/client/rpc-spec-v2/src/archive/archive_storage.rs

```diff
@@ -127,3 +139,724 @@ where
 		ArchiveStorageResult::ok(storage_results, discarded_items)
 	}
 }
+
+/// Parse hex-encoded string parameter as raw bytes.
+///
+/// If the parsing fails, returns an error propagated to the RPC method.
+pub fn parse_hex_param(param: String) -> Result<Vec<u8>, ArchiveError> {
+	// Methods can accept empty parameters.
+	if param.is_empty() {
+		return Ok(Default::default())
+	}
+
+	array_bytes::hex2bytes(&param).map_err(|_| ArchiveError::InvalidParam(param))
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct DiffDetails {
+	key: StorageKey,
+	return_type: ArchiveStorageDiffType,
+	child_trie_key: Option<ChildInfo>,
+	child_trie_key_string: Option<String>,
+}
+
+/// The type of storage query.
+#[derive(Debug, PartialEq, Clone, Copy)]
+enum FetchStorageType {
+	/// Only fetch the value.
+	Value,
+	/// Only fetch the hash.
+	Hash,
+	/// Fetch both the value and the hash.
+	Both,
+}
+
+pub struct ArchiveStorageDiff<Client, Block, BE> {
+	client: Storage<Client, Block, BE>,
+}
+
+impl<Client, Block, BE> ArchiveStorageDiff<Client, Block, BE> {
+	pub fn new(client: Arc<Client>) -> Self {
+		Self { client: Storage::new(client) }
+	}
+}
+
+impl<Client, Block, BE> ArchiveStorageDiff<Client, Block, BE>
+where
+	Block: BlockT + 'static,
+	BE: Backend<Block> + 'static,
+	Client: StorageProvider<Block, BE> + Send + Sync + 'static,
+{
+	/// Fetch the storage from the given key.
+	///
+	/// This method returns:
+	/// - `None` if the storage is not present.
+	/// - `Some((StorageResult, None))` for `FetchStorageType::Value`.
+	/// - `Some((StorageResult, None))` for `FetchStorageType::Hash`.
+	/// - `Some((StorageResult, Some(StorageResult)))` for `FetchStorageType::Both`.
+	fn fetch_storage(
+		&self,
+		hash: Block::Hash,
+		key: StorageKey,
+		maybe_child_trie: Option<ChildInfo>,
+		ty: FetchStorageType,
+	) -> Result<Option<(StorageResult, Option<StorageResult>)>, String> {
+		match ty {
+			FetchStorageType::Value => {
+				let result = self.client.query_value(hash, &key, maybe_child_trie.as_ref())?;
+
+				Ok(result.map(|res| (res, None)))
+			},
+
+			FetchStorageType::Hash => {
+				let result = self.client.query_hash(hash, &key, maybe_child_trie.as_ref())?;
+
+				Ok(result.map(|res| (res, None)))
+			},
+
+			FetchStorageType::Both => {
+				let value = self.client.query_value(hash, &key, maybe_child_trie.as_ref())?;
+				let Some(value) = value else {
+					return Ok(None);
+				};
+
+				let hash = self.client.query_hash(hash, &key, maybe_child_trie.as_ref())?;
+				let Some(hash) = hash else {
+					return Ok(None);
+				};
+
+				Ok(Some((value, Some(hash))))
+			},
+		}
+	}
+
+	/// Check if the key starts with any of the provided items.
+	///
```

Comment:
Perhaps add something that explains that if `items` is empty then the storage difference is calculated for all keys of the main storage trie.

For readability you could also remove the `is_empty stuff` because it's a bit hard to understand and check it explicitly before calling `starts_with`

Reply:
yeah it's better but even better would be `as_fetch_storage` or something like that but this is good enough :)